protocols/autonat/src/behaviour.rs

```diff
@@ -0,0 +1,853 @@
+        conn: &ConnectionId,
+        old: &ConnectedPoint,
+        new: &ConnectedPoint,
+    ) {
+        self.inner.inject_address_change(peer, conn, old, new);
+
+        self.connected
+            .insert(*peer, new.get_remote_address().clone());
+    }
+
+    fn inject_new_listen_addr(&mut self, id: ListenerId, addr: &Multiaddr) {
+        self.inner.inject_new_listen_addr(id, addr);
+
+        // New address could be publicly reachable, trigger retry.
+        if !self.nat_status.is_public() {
+            if self.confidence > 0 {
+                self.confidence -= 1;
+            }
+            self.schedule_next_probe(self.config.retry_interval);
+        }
+    }
+
+    fn inject_expired_listen_addr(&mut self, id: ListenerId, addr: &Multiaddr) {
+        self.inner.inject_expired_listen_addr(id, addr);
+        if let Some(public_address) = self.public_address() {
+            if public_address == addr {
+                self.confidence = 0;
+                self.nat_status = NatStatus::Unknown;
+                self.schedule_probe.reset(Duration::ZERO);
+            }
+        }
+    }
+
+    fn inject_new_external_addr(&mut self, addr: &Multiaddr) {
+        self.inner.inject_new_external_addr(addr);
+
+        // New address could be publicly reachable, trigger retry.
+        if !self.nat_status.is_public() {
+            if self.confidence > 0 {
+                self.confidence -= 1;
+            }
+            self.schedule_next_probe(self.config.retry_interval);
+        }
+    }
+
+    fn inject_expired_external_addr(&mut self, addr: &Multiaddr) {
+        self.inner.inject_expired_external_addr(addr);
+        if let Some(public_address) = self.public_address() {
+            if public_address == addr {
+                self.confidence = 0;
+                self.nat_status = NatStatus::Unknown;
+                self.schedule_probe.reset(Duration::ZERO);
+            }
+        }
+    }
+
+    fn poll(
+        &mut self,
+        cx: &mut Context<'_>,
+        params: &mut impl PollParameters,
+    ) -> Poll<NetworkBehaviourAction<Self::OutEvent, Self::ProtocolsHandler>> {
+        let mut is_probe_ready = false;
+        loop {
+            if self.schedule_probe.poll_unpin(cx).is_ready() {
+                is_probe_ready = true;
+                self.schedule_probe.reset(self.config.retry_interval);
+                continue;
+            }
+            if is_probe_ready {
+                let mut addresses: Vec<_> = params.external_addresses().map(|r| r.addr).collect();
+                addresses.extend(params.listened_addresses());
+
+                match self.do_probe(addresses) {
+                    Some(_) => {}
+                    None => {
+                        let nat_status = NatStatus::Unknown;
+                        let has_flipped = self.handle_reported_status(nat_status);
+                        if has_flipped {
+                            self.pending_out_events.push_back(self.nat_status.clone());
```

Comment:
I think I did it this way because a previous iteration of the loop could have also pushed an event to the queue, and I want to keep the correct order.